File: SWE/SWE_Zusammenfassung/java/src/main/java/javademo/DerivedClass.java
package javademo;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.NoSuchElementException;

public class DerivedClass<T> extends BaseClass<T> implements Iterable<T> {
    private ArrayList<T> data;
    private int currentIndex;

    public DerivedClass(int size) {
        super(size);
        data = new ArrayList<>(size); // generic arrays in Java are semi-supported
        currentIndex = 0;
    }

    @Override
    public T getValue(int index) {
        if (index < 0 || index >= getSize()) {
            throw new IndexOutOfBoundsException();
        }

        return data.get(index);
    }

    @Override
    public void addValue(T value) {
        if (currentIndex < 0 || currentIndex >= getSize()) {
            throw new IndexOutOfBoundsException();
        }

        data.add(value);
        currentIndex++;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < currentIndex; i++) {
            sb.append(data.get(i));
            sb.append(" ");
        }

        return sb.toString().trim();
    }

    // Iterable implementation for for-each loop
    @Override
    public Iterator<T> iterator() {
        return new Iterator<T>() {
            private int index = 0;

            @Override
            public boolean hasNext() {
                return index < currentIndex;
            }

            @Override
            public T next() {
                if (!hasNext()) {
                    throw new NoSuchElementException();
                }

                return data.get(index++);
            }
        };
    }

    // No operator overloading in Java, so no operator[] or operator+=
    // No need for destructors or manual memory management
}

File: SWE/SWE_Zusammenfassung/csharp/DerivedClass.cs
using System.Collections;
using System.Text;

namespace demo;

// IEnumerable is commonly used for iteration in C#, but it is not required because forEach can be used with any class that implements GetEnumerator(Duck Typing).
public class DerivedClass<T> : BaseClass<T>, IEnumerable<T>
{
    private readonly T[] data;
    private int currentIndex;

    // could also be a primary constructor like in the base class
    public DerivedClass(int size) : base(size)
    {
        data = new T[size];
        currentIndex = 0;
    }

    public override T GetValue(int index)
    {
        if (index < 0 || index >= Size)
        {
            throw new IndexOutOfRangeException();
        }

        return data[index];
    }

    public override void AddValue(T value)
    {
        if (currentIndex < 0 || currentIndex >= Size)
        {
            throw new IndexOutOfRangeException();
        }

        data[currentIndex++] = value;
    }

    // C# supports indexers
    public override T this[int index]
    {
        get { return GetValue(index); }
    }

    public static DerivedClass<T> operator +(DerivedClass<T> obj, T value)
    {
        obj.AddValue(value);
        return obj;
    }

    public override string ToString()
    {
        var stringBuilder = new StringBuilder();

        for (int i = 0; i < currentIndex; i++)
        {
            stringBuilder.Append($"{data[i]} ");
        }

        return stringBuilder.ToString().Trim();
    }

    public IEnumerator<T> GetEnumerator()
    {
        for (int i = 0; i < currentIndex; i++)
        {
            yield return data[i];
        }
    }

    // The non-generic version of GetEnumerator needs to be implemented explicitly
    IEnumerator IEnumerable.GetEnumerator()
    {
        return GetEnumerator();
    }

    // No need for destructors or manual memory management
}
